### PR TITLE
Xml comments support in synapse configuration

### PR DESCRIPTION
**Pull request by jagathsisira**
With this updated, selected XML configurations will preserve XML comments in ESB configuration files.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/synapse/config/xml/SynapseXMLConfigurationFactory.java [hunk 1]
```diff
@@ -99,11 +100,11 @@ public SynapseConfiguration getConfiguration(OMElement definitions, Properties p
                     defineEventSource(config, elt, properties);
                 } else if (XMLConfigConstants.EXECUTOR_ELT.equals(elt.getQName())) {
                     defineExecutor(config, elt, properties);
-                } else if(XMLConfigConstants.MESSAGE_STORE_ELT.equals(elt.getQName())) {
+                } else if (XMLConfigConstants.MESSAGE_STORE_ELT.equals(elt.getQName())) {
                     defineMessageStore(config, elt, properties);
```
Comment (isudana) [inline modules/core/src/main/java/org/apache/synapse/config/xml/SynapseXMLConfigurationFactory.java]: It's better if we do not include unrelated code changes like formatting etc. to this PR.

File: modules/core/src/main/java/org/apache/synapse/config/SynapseConfiguration.java [hunk 2]
```diff
@@ -2127,4 +2132,25 @@ private String getTenantDomain(SynapseEnvironment synapseEnvironment) {
         return null;
     }
 
+    /**
+     * Returns CommentedText List
+     *
+     * @return ArrayList of String contains commented text segments
+     */
+    public List<String> getCommentedTextList() {
```
Comment (prabathabey) [inline modules/core/src/main/java/org/apache/synapse/config/SynapseConfiguration.java]: Maybe it's a good idea not to expose internals such as the data structures used internally, etc as part of "method names".

File: modules/core/src/main/java/org/apache/synapse/config/SynapseConfiguration.java [hunk 3]
```diff
@@ -2127,4 +2132,25 @@ private String getTenantDomain(SynapseEnvironment synapseEnvironment) {
...
+        return commentedTextList;
+    }
+
+    /**
+     * Add new comment text entry to the existing list*
+     *
+     * @param comment String comment
+     */
+    public void addToCommentedTextList(String comment) {
```
Comment (prabathabey) [inline modules/core/src/main/java/org/apache/synapse/config/SynapseConfiguration.java]: Same goes here as well.

File: modules/core/src/main/java/org/apache/synapse/config/xml/AbstractListMediatorFactory.java [hunk 4]
```diff
@@ -34,18 +37,45 @@
...
+    }
+
+    /**
+     * Find and add all comment nodes to the parent mediator as Comment Mediators
+     *
+     * @param el OMElement to extract OMComment Nodes
+     * @param m  Mediator to be updated with extracted OMComment nodes
+     */
+    protected static void addAllCommentChildrenToMediator(OMElement el, ListMediator m) {
+        Iterator it = el.getChildren();
```
Comment (prabathabey) [inline modules/core/src/main/java/org/apache/synapse/config/xml/AbstractListMediatorFactory.java]: This particular iterator's scope is not properly restricted. Imagine someone using the same iterator instance somewhere down the code. If the same is modified within the while clause, then that can lead to inconsistent behaviours.

File: modules/core/src/main/java/org/apache/synapse/config/xml/AbstractMediatorFactory.java [hunk 5]
```diff
@@ -212,4 +215,24 @@ protected void handleException(String message) {
...
     }
+
+    /**
+     * Find and add all comment nodes to the mediator as a list
+     *
+     * @param el          OMElement to extract OMComment Nodes
+     * @param commentList ArrayList to be updated with extracted Comment Strings
+     */
+    protected static void addAllCommentChildrenToList(OMElement el, List<String> commentList) {
+        Iterator it = el.getChildren();
```
Comment (prabathabey) [inline modules/core/src/main/java/org/apache/synapse/config/xml/AbstractMediatorFactory.java]: Same goes here as well.

File: modules/core/src/main/java/org/apache/synapse/config/xml/CommentMediatorFactory.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+import javax.xml.namespace.QName;
+import java.util.Properties;
+
+/**
+ * Factory for Comment Mediators.
+ */
+public class CommentMediatorFactory extends AbstractMediatorFactory {
+
+    private static final QName Comment_Q = new QName(XMLConfigConstants.SYNAPSE_NAMESPACE, "comment");
```
Comment (prabathabey) [inline modules/core/src/main/java/org/apache/synapse/config/xml/CommentMediatorFactory.java]: Variable naming conventions are not properly followed.
